### DISCUSSION TITLE
Generating Replacements for Maven Dependencies

### DISCUSSION THREAD

**Discussion by lars-sh** [renovatebot/renovate#30133] [Suggest an Idea]
### Tell us more.

I wondered how to generate replacement rules for Maven dependencies in a convenient way. Since my job is based on a small shell script I came up with the following solution that I'd like to share with you. I'm open for feedback.
Probably it helps someone else too or can be integrated into Renovate at a later point in time.

At first I'm downloading the files from https://github.com/jonathanlermitage/oga-maven-plugin/tree/master/uc.
Those files are maintained as part of the [Old GroupIds Alerter (OGA) Maven Plugin](https://github.com/jonathanlermitage/oga-maven-plugin).

Next I execute JQ to transform the definitions file to valid Renovate `packageRules`.
Here's the JQ script I'm using:
```
[
	.migration[]

	# Ignore removed artifacts
	| select(.new != "removed")
	| select(.new != "removed:removed")

	# Ignore artifacts with multiple new group names
	| select(.new | contains(" ") | not)
	| select(.new | contains("{") | not)

	# Create the Renovate JSON structure
	|
		# for single artifacts
		if .new | contains(":") then
			{
				description: .context,
				matchManagers: ["maven"],
				matchPackageNames: [.old],
				replacementName: .new
			}
		# for all artifacts with the same group name
		else
			{
				description: .context,
				matchManagers: ["maven"],
				matchPackagePrefixes: [.old + ":"],
				replacementNameTemplate: ("{{{replace '" + .old + ":' '" + .new + ":' packageName}}}")
			}
		end

	# Remove the description attribute if null
	| if .description == null then del(.description) else . end
]
```

For the unofficial dependencies the JQ script is slightly different:
```
[
	.migration[]

	# Ignore artifacts with more than one proposal
	| select(.proposal | length == 1)

	# Create the Renovate JSON structure
	|
		{
			description: .context,
			matchManagers: ["maven"],
			matchPackageNames: [.old],
			replacementName: .proposal[0]
		}

	# Remove the description attribute if null
	| if .description == null then del(.description) else . end
]
```

PS: For now Renovate warns on Maven dependencies with existing replacements. Support for replacement PRs is discussed at #24883.

**Comment by rarkins** [comment]
This is an interesting idea. 

If we want this to be part of default replacements then we'd want it to be built into Renovate so that it doesn't require an API call to download each time, and so that it doesn't ever just disappear or get wiped.

We'd also want to have a sanity check review of all changes to make sure nobody added in replacements which make no sense (including replacing a commonly used package with a malicious one).

That folder has had 12 commits this year, for reference. So we could feasibly even run/update this manually, or have a scheduled github action which regenerates the file and submits a PR if the result has changed.
